File: packages/agent/src/infrastructure/events/types/session.ts
/**
 * @fileoverview Session Lifecycle Events
 *
 * Events for session start, end, fork, and branch operations.
 */

import type { EventId, SessionId, BranchId } from './branded.js';
import type { BaseEvent } from './base.js';
import type { TokenUsage } from './token-usage.js';

// =============================================================================
// Session Events
// =============================================================================

/**
 * Session start event - root of a session tree
 */
export interface SessionStartEvent extends BaseEvent {
  type: 'session.start';
  payload: {
    workingDirectory: string;
    model: string;
    /** Provider (optional - can be auto-detected from model name) */
    provider?: string;
    systemPrompt?: string;
    title?: string;
    tags?: string[];
    /** If this is a fork, reference the source */
    forkedFrom?: {
      sessionId: SessionId;
      eventId: EventId;
    };
  };
}

/**
 * Session end event
 */
export interface SessionEndEvent extends BaseEvent {
  type: 'session.end';
  payload: {
    reason: 'completed' | 'aborted' | 'error' | 'timeout';
    summary?: string;
    totalTokenUsage?: TokenUsage;
    duration?: number; // milliseconds
  };
}

/**
 * Session fork event - marks a fork point
 */
export interface SessionForkEvent extends BaseEvent {
  type: 'session.fork';
  payload: {
    /** Source session we forked from */
    sourceSessionId: SessionId;
    /** Event ID we forked from */
    sourceEventId: EventId;
    /** Name for the fork */
    name?: string;
    /** Why was this forked */
    reason?: string;
  };
}

/**
 * Named branch creation
 */
export interface SessionBranchEvent extends BaseEvent {
  type: 'session.branch';
  payload: {
    branchId: BranchId;
    name: string;
    description?: string;
  };
}

File: packages/agent-rs/crates/tron-events/src/types/payloads/session.rs
//! Session lifecycle payloads: start, end, fork.

use serde::{Deserialize, Serialize};

/// Payload for `session.start` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionStartPayload {
    /// Absolute path to the working directory.
    pub working_directory: String,
    /// LLM model ID.
    pub model: String,
    /// Provider name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    /// System prompt content.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub system_prompt: Option<String>,
    /// Session title.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    /// Session tags.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    /// Fork source information.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub forked_from: Option<ForkSource>,
}

/// Fork source reference.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ForkSource {
    /// Source session ID.
    pub session_id: String,
    /// Source event ID (fork point).
    pub event_id: String,
}

/// Payload for `session.end` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionEndPayload {
    /// End reason.
    pub reason: String,
    /// Session summary.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
    /// Aggregate token usage.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total_token_usage: Option<super::TokenUsage>,
    /// Duration in milliseconds.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration: Option<i64>,
}

/// Payload for `session.fork` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionForkPayload {
    /// Session being forked from.
    pub source_session_id: String,
    /// Event ID at the fork point.
    pub source_event_id: String,
    /// Fork name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// Reason for fork.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}
